### PR TITLE
generalize select batch rule (fixes #311)

### LINKED ISSUE #311: dev timeline for more batched ops
Hi JAX Team,

I'm interested in using your library, e.g. to use gradients to optimize functions whose
terms include functions g() of jacobians of functions f(). There's no good way in PyTorch to do this for f() with output dim > 1, and certainly no good way to do this in a batched setting.

In JAX, i'm able to very easily get e.g. the trace of the jacobian of a plain relu neural net. So cool! Things break down when I try to use vmap to get a batched version of the computation.  I'm wondering, do you have a rough guess at the timeline for when I might be able to vmap in this kind of setting? Note: I've successfully used JAX to do vmap's of functions g() of jacobians of functions f() for simple choices of f()

```python
import jax.numpy as np
from jax import jit, jacrev, vmap

W = np.eye(3)
b = np.zeros(3)

def relu(x):
    return np.maximum(0,x)

def NN(x):
    return relu(np.dot(W,x) + b)

Jx_NN = jit(jacrev(NN))

# By the way, is there a recommended way to compose already JIT'ed functions?
def trace_Jx_NN(x):
    J = Jx_NN(x)
    return np.trace(J) 


x1 = np.array([2.0,2.0,2.0])
x2 = np.array([3.0,3.0,3.0])
X = np.vstack((x1,x2))

# this works
print(trace_Jx_NN(x1)) 

# this line executes
batched_trace_Jx_NN = jit(vmap(trace_Jx_NN,in_axes=(0)))

# raises "NotImplementedError  # TODO(schsam, mattjj): Handle more cases."
print(batched_trace_Jx_NN(X)) 
```

Thanks!
mark

### PR DESCRIPTION
**Pull request by mattjj**

### CODE REVIEW

**Review by froystig — APPROVED** [review]

File: jax/lax.py [hunk 1]
```diff
@@ -1755,17 +1755,30 @@ def _select_transpose_rule(t, pred, on_true, on_false):
             select(pred, zeros, t) if on_false is None else None]
 
 def _select_batch_rule(batched_args, batch_dims, **unused_kwargs):
```
Comment (hawkinsp) [inline jax/lax.py]: (I haven't studied this change carefully yet.)

Would this rule be simplified if we generalized the semantics of `lax.select` to support more general broadcasting? 

The XLA `Select` semantics are a little awkward (e.g., only scalar broadcasting) so I wrote a wrapper that does general broadcasting, not just scalar/non-scalar:
https://github.com/google/jax/blob/master/jax/lax.py#L1001

Comment (mattjj) [reply jax/lax.py]: Yes, definitely. That's great to know about! These commits could probably be simplified by using that helper function.
